src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/Authenticator.cs

```diff
@@ -122,10 +125,35 @@ public async Task UpdateFromTemplateAsync(RequestContext requestContext)
                 {
                     ServiceBundle.InstanceDiscovery.AddMetadataEntry(host);
                 }
-                AuthorizationUri = ServiceBundle.InstanceDiscovery.FormatAuthorizeEndpoint(host, tenant);
-                DeviceCodeUri = string.Format(CultureInfo.InvariantCulture, "https://{0}/{1}/oauth2/devicecode", host, tenant);
-                TokenUri = string.Format(CultureInfo.InvariantCulture, "https://{0}/{1}/oauth2/token", host, tenant);
-                UserRealmUriPrefix = EnsureUrlEndsWithForwardSlash(string.Format(CultureInfo.InvariantCulture, "https://{0}/common/userrealm", host));
+
+                //Check if custom port is set on authority
+                if (authorityUri.Port != DefaultHttpsPort)
```

Comment:
So the logic here is good but it is too repetitive, i.e. you write the same code 4 times. How about creating a static helper in `CoreHelpers` or `UriBuilderExtensions`

`UriBuilder GetHttpsUriWithOptionalPort(string host, string tenant, string path, int port)`

The advantages of having this are: 

- very easy to test (your existing test is good, but you can write more tests this way)
- very easy to move to MSAL, along with tests